File: android/app/src/main/kotlin/org/recall/mic/Handshake.kt
package org.recall.mic

import java.util.Locale

// The ingest handshake line — pure string building, so it's unit-tested on the
// JVM (the streaming service just calls it).

/**
 * The one-line announcement sent on the shared ingest port before any PCM:
 * identity, stream shape, and the capture epoch of the first PCM byte.
 *
 * `epochMillis` is the phone's wall-clock right after AudioRecord.startRecording()
 * — the capture instant of the earliest samples the first read will drain. The
 * server measures that byte's arrival and renames this connection's segments from
 * arrival time back to capture time, so cross-mic timestamps share a clock.
 * Rendered as fixed-point seconds (never Double interpolation, which turns large
 * values into scientific notation), Locale.ROOT so the decimal point survives
 * every locale.
 */
fun handshakeLine(deviceId: String, sampleRate: Int, epochMillis: Long): String {
    val epoch = String.format(Locale.ROOT, "%d.%03d", epochMillis / 1000, epochMillis % 1000)
    return """{"id":"$deviceId","rate":$sampleRate,"channels":1,"epoch":$epoch}""" + "\n"
}

File: ios/Sources/Handshake.swift
import Foundation

/// The ingest handshake line — a direct port of the Android app's `Handshake.kt`,
/// kept pure so the two announcements can't drift apart and the format is testable
/// without a socket.
enum Handshake {
    /// The one-line announcement sent on the shared ingest port before any PCM:
    /// identity, stream shape, and the capture epoch of the first PCM byte.
    ///
    /// `epoch` is the wall-clock at handshake time. The capture engine streams
    /// continuously and PCM is dropped while disconnected, so the first block a
    /// fresh connection forwards was captured essentially now — within one tap
    /// buffer, and always at or before its arrival at the server, which is the
    /// side the server's clamp accepts. The server measures that first byte's
    /// arrival and renames this connection's segments from arrival time back to
    /// capture time, so cross-mic timestamps share a clock.
    ///
    /// Field order and the trailing newline must match the recorder's parser
    /// exactly; the epoch is fixed-point (never exponent notation) with a POSIX
    /// locale so the decimal point survives every region setting.
    static func line(id: String, rate: Int, epoch: TimeInterval) -> Data {
        let stamp = String(format: "%.3f", locale: Locale(identifier: "en_US_POSIX"), epoch)
        return Data("{\"id\":\"\(id)\",\"rate\":\(rate),\"channels\":1,\"epoch\":\(stamp)}\n".utf8)
    }
}
